File: prob-model/plotResults.m
% SERIAL
load experimentResultsSerial
% Baseline
[max_K, N] = size(baseline_correctness_serial);
% determine how many entries have actually been computed
N = max([1:N](logical(sum(baseline_correctness_serial))));
% transform the prediction hits/misses into a rates incrementally
% clustering-based
% K x N x accuracy/precision/recall/F-measure
baseline_hit_rate = zeros(max_K, N, 4);
for K = [2:max_K]
  for n = 2:N
    % Accuracy
    baseline_hit_rate(K, n, 1) = sum(baseline_correctness_serial(K, 2:n))/(n - 1);
    % Precision
    Z = sum(baseline_correctness_serial(K, ~logical(d(2:n)))) + sum(!baseline_correctness_serial(K, logical(d(2:n))));
    if(Z > 0)
      baseline_hit_rate(K, n, 2) = sum(baseline_correctness_serial(K, ~logical(d(2:n)))) / Z;
    else
      baseline_hit_rate(K, n, 2) = 0;
    endif
    % Recall
    if(sum(!d(2:n)) > 0)
      baseline_hit_rate(K, n, 3) = sum(baseline_correctness_serial(K, ~logical(d(2:n)))) / sum(!d(2:n));
    else
      baseline_hit_rate(K, n, 3) = 0;
    endif
    % F-measure
    Z = baseline_hit_rate(K, n, 2) + baseline_hit_rate(K, n, 3);
    if(Z > 0)
      baseline_hit_rate(K, n, 4) = 2*baseline_hit_rate(K, n, 2)*baseline_hit_rate(K, n, 3)/Z;
    else
      baseline_hit_rate(K, n, 4) = 0;
    endif
  endfor
endfor
% Prob model
[max_K, N] = size(prob_model_correctness_serial);
% determine how many entries have actually been computed
N = max([1:N](logical(sum(prob_model_correctness_serial))));
% transform the prediction hits/misses into a rates incrementally
% clustering-based
% K x N x accuracy/precision/recall/F-measure
prob_model_hit_rate = zeros(max_K, N, 4);
for K = [2:max_K]
  for n = 2:N
    % Accuracy
    prob_model_hit_rate(K, n, 1) = sum(prob_model_correctness_serial(K, 2:n))/(n - 1);
    % Precision
    Z = sum(prob_model_correctness_serial(K, ~logical(d(2:n)))) + sum(!prob_model_correctness_serial(K, logical(d(2:n))));
    if(Z > 0)
      prob_model_hit_rate(K, n, 2) = sum(prob_model_correctness_serial(K, ~logical(d(2:n)))) / Z;
    else
      prob_model_hit_rate(K, n, 2) = 0;
    endif
    % Recall
    if(sum(!d(2:n)) > 0)
      prob_model_hit_rate(K, n, 3) = sum(prob_model_correctness_serial(K, ~logical(d(2:n)))) / sum(!d(2:n));
    else
      prob_model_hit_rate(K, n, 3) = 0;
    endif
    % F-measure
    Z = prob_model_hit_rate(K, n, 2) + prob_model_hit_rate(K, n, 3);
    if(Z > 0)
      prob_model_hit_rate(K, n, 4) = 2*prob_model_hit_rate(K, n, 2)*prob_model_hit_rate(K, n, 3)/Z;
    else
      prob_model_hit_rate(K, n, 4) = 0;
    endif
  endfor
endfor
% SVM
% total rate/precision/recall/F-measure
svm_hit_rate = zeros(4, N);
for n = 2:N
  % get the values that count at all
  valid_idxs = svm_correctness_serial(2:n) > -1;
  % count the hits
  svm_hits = svm_correctness_serial(2:n) .* double(valid_idxs);
  if(sum(double(valid_idxs)) > 0)
    % Accuracy
    svm_hit_rate(1, n) = sum(svm_hits)/(n - 1);
    %svm_hit_rate(1, n) = sum(svm_hits)/sum(double(valid_idxs));
    % Precision
    Z = sum(svm_hits .* !d(2:n)) + sum(!svm_hits(logical(d(2:n))));
    if(Z > 0)
      svm_hit_rate(2, n) = sum(svm_hits .* !d(2:n))/Z;
    else
      svm_hit_rate(2, n) = 0;
    endif
    % Recall
    Z = sum(svm_hits .* !d(2:n)) + sum(!svm_hits(logical(!d(2:n))));
    if(Z > 0)
      svm_hit_rate(3, n) = sum(svm_hits .* !d(2:n))/Z;
    else
      svm_hit_rate(3, n) = 0;
    endif
    % F-measure
    Z = svm_hit_rate(2, n) + svm_hit_rate(3, n);
    if(Z > 0)
      svm_hit_rate(4, n) = 2*svm_hit_rate(2, n)*svm_hit_rate(3, n)/Z;
    endif
  else
    svm_hit_rate(1, n) = 0;
    svm_hit_rate(2, n) = 0;
    svm_hit_rate(3, n) = 0;
    svm_hit_rate(4, n) = 0;
  endif
endfor
% Bernoulli
% total rate/precision/recall/F-measure
bernoulli_hit_rate = zeros(4, N);
for n = 2:N
  % get the values that count at all
  % count the hits
  bernoulli_hits = bernoulli_correctness_serial(2:n);
  % Accuracy
  bernoulli_hit_rate(1, n) = sum(bernoulli_hits)/(n - 1);
  % Precision
  Z = sum(bernoulli_hits .* !d(2:n)) + sum(!bernoulli_hits(logical(d(2:n))));
  if(Z > 0)
    bernoulli_hit_rate(2, n) = sum(bernoulli_hits .* !d(2:n))/Z;
  else
    bernoulli_hit_rate(2, n) = 0;
  endif
  % Recall
  Z = sum(bernoulli_hits .* !d(2:n)) + sum(!bernoulli_hits(logical(!d(2:n))));
  if(Z > 0)
    bernoulli_hit_rate(3, n) = sum(bernoulli_hits .* !d(2:n))/Z;
  else
    bernoulli_hit_rate(3, n) = 0;
  endif
  % F-measure
  Z = bernoulli_hit_rate(2, n) + bernoulli_hit_rate(3, n);
  if(Z > 0)
    bernoulli_hit_rate(4, n) = 2*bernoulli_hit_rate(2, n)*bernoulli_hit_rate(3, n)/Z;
  endif
endfor
% hit rate - prob_model
figure()
for K = 2:max_K
  % total rate
  % subplot(max_K - 1, 3, 3*(K - 2) + 1)
  subplot(max_K - 1, 1, K - 1)
  plot([2:N], prob_model_hit_rate(K, 2:N, 1), ';Accuracy;', prob_model_hit_rate(K, 2:N, 2), ';Precision;', prob_model_hit_rate(K, 2:N, 3), ';Recall;', prob_model_hit_rate(K, 2:N, 4), ';F-measure;');
  ylabel(['Baseline accuracy, K = ', num2str(K)]);
  %% 1-rate
  %subplot(max_K - 1, 3, 3*(K - 2) + 2)
  %plot([2:N], prob_model_hit_rate(2, 2:N, 2), ';1-rate;');
  %% 0-rate
  %subplot(max_K - 1, 3, 3*(K - 2) + 3)
  %plot([2:N], prob_model_hit_rate(2, 2:N, 3), ';0-rate;');
endfor
% hit rate - baseline
figure()
for K = 2:max_K
  % total rate
  % subplot(max_K - 1, 3, 3*(K - 2) + 1)
  subplot(max_K - 1, 1, K - 1)
  plot([2:N], baseline_hit_rate(K, 2:N, 1), ';Accuracy;', baseline_hit_rate(K, 2:N, 2), ';Precision;', baseline_hit_rate(K, 2:N, 3), ';Recall;', baseline_hit_rate(K, 2:N, 4), ';F-measure;');
  ylabel(['Baseline accuracy, K = ', num2str(K)]);
  %% 1-rate
  %subplot(max_K - 1, 3, 3*(K - 2) + 2)
  %plot([2:N], baseline_hit_rate(2, 2:N, 2), ';1-rate;');
  %% 0-rate
  %subplot(max_K - 1, 3, 3*(K - 2) + 3)
  %plot([2:N], baseline_hit_rate(2, 2:N, 3), ';0-rate;');
endfor
% hit rate - SVM
figure()
plot([2:N], svm_hit_rate(1, 2:N), ';Accuracy;', svm_hit_rate(2, 2:N), ';Precision;', svm_hit_rate(3, 2:N), ';Recall;', svm_hit_rate(4, 2:N), ';F-measure;');
ylabel('SVM accuracy');
% hit rate - Bernoulli
figure()
plot([2:N], bernoulli_hit_rate(1, 2:N), ';Accuracy;', bernoulli_hit_rate(2, 2:N), ';Precision;', bernoulli_hit_rate(3, 2:N), ';Recall;', bernoulli_hit_rate(4, 2:N), ';F-measure;');
ylabel('Bernoulli accuracy');
% learning - prob_model
if(length(prob_model_training_serial(K, :)) > N)
  prob_model_training_serial(:, N+1:end) = [];
endif
figure()
for K = [2:max_K]
  subplot(max_K - 1, 1, K - 1)
  size(prob_model_training_serial(K, 2:end))
  [2:N]
  N
  plot([2:N], prob_model_training_serial(K, 2:end), ';Learning time;');
  ylabel(['K = ', num2str(K)]);
endfor
% learning - baseline
if(length(baseline_training_serial(K, :)) > N)
  baseline_training_serial(:, N+1:end) = [];
endif
figure()
for K = [2:max_K]
  subplot(max_K - 1, 1, K - 1)
  size(baseline_training_serial(K, 2:end))
  [2:N]
  plot([2:N], baseline_training_serial(K, 2:end), ';Learning time;');
  ylabel(['K = ', num2str(K)]);
endfor
% learning - SVM
if(length(svm_training_serial) > N)
  svm_training_serial(N+1:end) = [];
endif
% set the "invalid" values to 0
svm_training_serial(svm_training_serial == -1) = 0;
figure()
plot([2:N], svm_training_serial(2:N), ';Learning time;');
ylabel('SVM');
% learning - Bernoulli
if(length(bernoulli_training_serial) > N)
  bernoulli_training_serial(N+1:end) = [];
endif
figure()
plot([2:N], bernoulli_training_serial(2:N), ';Learning time;');
ylabel('Bernoulli');
% prediction - prob_model
if(length(prob_model_prediction_serial(K, :)) > N)
  prob_model_prediction_serial(:, N+1:end) = [];
endif
figure()
for K = [2:max_K]
  subplot(max_K - 1, 1, K - 1)
  plot([2:N], prob_model_prediction_serial(K, 2:N), ';Prediction time;');
  ylabel(['K = ', num2str(K)]);
endfor
% prediction - baseline
if(length(baseline_prediction_serial(K, :)) > N)
  baseline_prediction_serial(:, N+1:end) = [];
endif
figure()
for K = [2:max_K]
  subplot(max_K - 1, 1, K - 1)
  plot([2:N], baseline_prediction_serial(K, 2:N), ';Prediction time;');
  ylabel(['K = ', num2str(K)]);
endfor
% prediction - SVM
if(length(svm_prediction_serial) > N)
  svm_prediction_serial(N+1:end) = [];
endif
% set the "invalid" values to 0
svm_prediction_serial(svm_prediction_serial == -1) = 0;
figure()
plot([2:N], svm_prediction_serial(2:N), ';Prediction time;');
ylabel('SVM');
% prediction - Bernoulli
if(length(bernoulli_prediction_serial) > N)
  bernoulli_prediction_serial(N+1:end) = [];
endif
figure()
plot([2:N], bernoulli_prediction_serial(2:N), ';Prediction time;');
ylabel('Bernoulli');

% save the processed data for plotting it in R
save prob_model_hit_rate_serial.mat prob_model_hit_rate 
save prob_model_training_serial.mat prob_model_training_serial
save prob_model_prediction_serial.mat prob_model_prediction_serial
save baseline_hit_rate_serial.mat baseline_hit_rate 
save baseline_training_serial.mat baseline_training_serial
save baseline_prediction_serial.mat baseline_prediction_serial
save svm_hit_rate.mat svm_hit_rate
save svm_training_serial.mat svm_training_serial
save svm_prediction_serial.mat svm_prediction_serial
save bernoulli_hit_rate.mat bernoulli_hit_rate
save bernoulli_training_serial.mat bernoulli_training_serial
save bernoulli_prediction_serial.mat bernoulli_prediction_serial
disp('The processed experimental results (serial) have been saved')

%% PARALLEL
%load experimentResultsParallel
%save baseline_accuracy.mat baseline_accuracy 
%save baseline_learning.mat baseline_learning 
%save baseline_prediction.mat baseline_prediction
%save prob_accuracy.mat prob_accuracy  
%save prob_learning.mat prob_learning  
%save prob_prediction.mat prob_prediction  
%save svm_accuracy.mat svm_accuracy  
%save svm_learning.mat svm_learning  
%save svm_prediction.mat svm_prediction
%disp('The processed experimental results (parallel) have been saved')

File: prob-model/plotting/plotResults.r
library("foreign")
X11.options(type="Xlib") 

# baseline hit rates for Ks separate
filename <- "baseline_hit_rate_serial";
baseline_hit_rate_serial <- read.octave(paste("serial/", filename, ".mat", sep=""))
baseline_hit_rate_serial <- baseline_hit_rate_serial[[1]]
max_K = length(baseline_hit_rate_serial[, 1, 1])
for(K in 2:max_K) {
  pdf(file=paste(filename, "-K", K, ".pdf", sep=""), width=7, height=4)
  last <- length(baseline_hit_rate_serial[K, , 1])
  plot(2:last, baseline_hit_rate_serial[K, 2:last, 1], type="l", col="black", ylim=c(0, 1), xlab="Iteration", ylab="Hit rate", lty=1, lwd=2)
  lines(2:last, baseline_hit_rate_serial[K, 2:last, 2], type="l", col="blue", pch=22, lty=2, lwd=2)
  lines(2:last, baseline_hit_rate_serial[K, 2:last, 3], type="l", col="red", pch=22, lty=3, lwd=2)
  lines(2:last, baseline_hit_rate_serial[K, 2:last, 4], type="l", col="green", pch=22, lty=4, lwd=2)
  title(main="Clustering Quality", col.main="black", font.main=4)
  legend(x="bottomright", legend=c("Accuracy", "Precision", "Recall", "F-measure"), col=c("black", "blue", "red", "green"), lty=1:3, lwd=2)
  dev.off()
}

## plot total rates for various K for comparison
#pdf(file=paste(filename, "-K-comparison", ".pdf", sep=""), width=7, height=4)
#last <- length(baseline_hit_rate_serial[2, , 1])
#plot(2:last, baseline_hit_rate_serial[2, 2:last, 1], type="l", col="black", ylim=c(0, 1), xlab="Iteration", ylab="Hit rate", lty=1, lwd=2)
#for(K in 3:max_K) {
#  lines(2:last, baseline_hit_rate_serial[K, 2:last, 1], type="l", col="black", lty=K-1, lwd=2)
#}
#title(main="Clustering Hit Rate", col.main="black", font.main=4)
#legend(x="bottomright", legend=c(2:max_K), lty=c(1:(max_K-1)), lwd=2)
#dev.off()

# learning run times
filename <- "baseline_training_serial";
baseline_training_serial <- read.octave(paste("serial/", filename, ".mat", sep=""))
baseline_training_serial <- baseline_training_serial[[1]]
max_K = length(baseline_training_serial[, 1])
last <- length(baseline_training_serial[2, ])
maxY <- max(baseline_training_serial)
for(K in 2:max_K) {
  pdf(file=paste(filename, "-K_", K, ".pdf", sep=""), width=7, height=4)
  plot(2:last, baseline_training_serial[K, 2:last], ylim=c(0, maxY), type="l", col="black", xlab="Iteration", ylab="Training time [s]", lty=1, lwd=2)
  title(main=paste("Clustering training times, K = ", K), col.main="black", font.main=4)
  dev.off()
}

# clustering prediction times
filename <- "baseline_prediction_serial";
baseline_prediction_serial <- read.octave(paste("serial/", filename, ".mat", sep=""))
baseline_prediction_serial <- baseline_prediction_serial[[1]]
max_K = length(baseline_prediction_serial[, 1])
last <- length(baseline_prediction_serial[2, ])
maxY <- max(baseline_prediction_serial)
for(K in 2:max_K) {
  pdf(file=paste(filename, "-K_", K, ".pdf", sep=""), width=7, height=4)
  plot(2:last, baseline_prediction_serial[K, 2:last], ylim=c(0, maxY), type="l", col="black", xlab="Iteration", ylab="Prediction time [s]", lty=1, lwd=2)
  title(main=paste("Clustering prediction times, K = ", K), col.main="black", font.main=4)
  dev.off()
}

# SVM hit rate
filename <- "svm_hit_rate";
svm_hit_rate <- read.octave(paste("serial/", filename, ".mat", sep=""))
svm_hit_rate <- svm_hit_rate[[1]]
pdf(file=paste(filename, ".pdf", sep=""), width=7, height=4)
last <- length(svm_hit_rate[1, ])
plot(2:last, svm_hit_rate[1, 2:last], type="l", col="black", ylim=c(0, 1), xlab="Iteration", ylab="Hit rate", lty=1, lwd=2)
lines(2:last, svm_hit_rate[2, 2:last], type="l", col="blue", pch=22, lty=2, lwd=2)
lines(2:last, svm_hit_rate[3, 2:last], type="l", col="red", pch=22, lty=3, lwd=2)
lines(2:last, svm_hit_rate[4, 2:last], type="l", col="green", pch=22, lty=4, lwd=2)
title(main="SVM Quality", col.main="black", font.main=4)
legend(x="bottomright", legend=c("Accuracy", "Precision", "Recall", "F-measure"), col=c("black", "blue", "red", "green"), lty=1:3, lwd=2)
dev.off()

# SVM training times
filename <- "svm_training_serial";
svm_training_serial <- read.octave(paste("serial/", filename, ".mat", sep=""))
svm_training_serial <- svm_training_serial[[1]]
last <- length(svm_training_serial)
pdf(file=paste(filename, ".pdf", sep=""), width=7, height=4)
plot(2:last, svm_training_serial[2:last], type="l", col="black", xlab="Iteration", ylab="Training time [s]", lty=1, lwd=2)
title(main="SVM training times", col.main="black", font.main=4)
dev.off()

# SVM prediction times
filename <- "svm_prediction_serial";
svm_prediction_serial <- read.octave(paste("serial/", filename, ".mat", sep=""))
svm_prediction_serial <- svm_prediction_serial[[1]]
last <- length(svm_prediction_serial)
pdf(file=paste(filename, ".pdf", sep=""), width=7, height=4)
plot(2:last, svm_prediction_serial[2:last], type="l", col="black", xlab="Iteration", ylab="Prediction time [s]", lty=1, lwd=2)
title(main="SVM prediction times", col.main="black", font.main=4)
dev.off()

# Bernoulli hit rate
filename <- "bernoulli_hit_rate";
bernoulli_hit_rate <- read.octave(paste("serial/", filename, ".mat", sep=""))
bernoulli_hit_rate <- bernoulli_hit_rate[[1]]
pdf(file=paste(filename, ".pdf", sep=""), width=7, height=4)
last <- length(bernoulli_hit_rate[1, ])
plot(2:last, bernoulli_hit_rate[1, 2:last], type="l", col="black", ylim=c(0, 1), xlab="Iteration", ylab="Hit rate", lty=1, lwd=2)
lines(2:last, bernoulli_hit_rate[2, 2:last], type="l", col="blue", pch=22, lty=2, lwd=2)
lines(2:last, bernoulli_hit_rate[3, 2:last], type="l", col="red", pch=22, lty=3, lwd=2)
lines(2:last, bernoulli_hit_rate[4, 2:last], type="l", col="green", pch=22, lty=4, lwd=2)
title(main="Bernoulli Quality", col.main="black", font.main=4)
legend(x="bottomright", legend=c("Accuracy", "Precision", "Recall", "F-measure"), col=c("black", "blue", "red", "green"), lty=1:3, lwd=2)
dev.off()

# Bernoulli training times
filename <- "bernoulli_training_serial";
bernoulli_training_serial <- read.octave(paste("serial/", filename, ".mat", sep=""))
bernoulli_training_serial <- bernoulli_training_serial[[1]]
last <- length(bernoulli_training_serial)
pdf(file=paste(filename, ".pdf", sep=""), width=7, height=4)
plot(2:last, bernoulli_training_serial[2:last], type="l", col="black", xlab="Iteration", ylab="Training time [s]", lty=1, lwd=2)
title(main="Bernoulli training times", col.main="black", font.main=4)
dev.off()

# Bernoulli prediction times
filename <- "bernoulli_prediction_serial";
bernoulli_prediction_serial <- read.octave(paste("serial/", filename, ".mat", sep=""))
bernoulli_prediction_serial <- bernoulli_prediction_serial[[1]]
last <- length(bernoulli_prediction_serial)
pdf(file=paste(filename, ".pdf", sep=""), width=7, height=4)
plot(2:last, bernoulli_prediction_serial[2:last], type="l", col="black", xlab="Iteration", ylab="Prediction time [s]", lty=1, lwd=2)
title(main="Bernoulli prediction times", col.main="black", font.main=4)
dev.off()

# SVM and clustering accuracy comparison
filename <- "svm_hit_rate";
svm_hit_rate <- read.octave(paste("serial/", filename, ".mat", sep=""))
svm_hit_rate <- svm_hit_rate[[1]]
filename <- "bernoulli_hit_rate";
bernoulli_hit_rate <- read.octave(paste("serial/", filename, ".mat", sep=""))
bernoulli_hit_rate <- bernoulli_hit_rate[[1]]
filename <- "baseline_hit_rate_serial";
baseline_hit_rate_serial <- read.octave(paste("serial/", filename, ".mat", sep=""))
baseline_hit_rate_serial <- baseline_hit_rate_serial[[1]]
pdf(file=paste("accuracy-comparison", ".pdf", sep=""), width=7, height=4)
last <- length(svm_hit_rate[1, ])
plot(2:last, svm_hit_rate[1, 2:last], type="l", col="black", ylim=c(0, 1), xlab="Iteration", ylab="Accuracy", lty=1, lwd=2)
last <- length(bernoulli_hit_rate[1, ])
lines(2:last, bernoulli_hit_rate[1, 2:last], type="l", col="red", ylim=c(0, 1), lty=2, lwd=2)
last <- length(baseline_hit_rate_serial[max_K, , 1])
best_K <- which.max(baseline_hit_rate_serial[, last, 1])
lines(2:last, baseline_hit_rate_serial[best_K, 2:last, 1], type="l", col="blue", ylim=c(0, 1), lty=3, lwd=2)
title(main="Accuracy Comparison", col.main="black", font.main=4)
legend(x="bottomright", legend=c("SVM", "Bernoulli", paste("Clustering, K=", best_K)), col=c("black", "red", "blue"), lty=c(1:3), lwd=2)
dev.off()

# SVM and clustering precision comparison
filename <- "svm_hit_rate";
svm_hit_rate <- read.octave(paste("serial/", filename, ".mat", sep=""))
svm_hit_rate <- svm_hit_rate[[1]]
filename <- "bernoulli_hit_rate";
bernoulli_hit_rate <- read.octave(paste("serial/", filename, ".mat", sep=""))
bernoulli_hit_rate <- bernoulli_hit_rate[[1]]
filename <- "baseline_hit_rate_serial";
baseline_hit_rate_serial <- read.octave(paste("serial/", filename, ".mat", sep=""))
baseline_hit_rate_serial <- baseline_hit_rate_serial[[1]]
pdf(file=paste("precision-comparison", ".pdf", sep=""), width=7, height=4)
last <- length(svm_hit_rate[1, ])
plot(2:last, svm_hit_rate[2, 2:last], type="l", col="black", ylim=c(0, 1), xlab="Iteration", ylab="Precision", lty=1, lwd=2)
last <- length(bernoulli_hit_rate[1, ])
lines(2:last, bernoulli_hit_rate[2, 2:last], type="l", col="red", ylim=c(0, 1), lty=2, lwd=2)
last <- length(baseline_hit_rate_serial[best_K, , 3])
lines(2:last, baseline_hit_rate_serial[best_K, 2:last, 2], type="l", col="blue", ylim=c(0, 1), lty=3, lwd=2)
title(main="Precision Comparison", col.main="black", font.main=4)
legend(x="bottomright", legend=c("SVM", "Bernoulli", paste("Clustering, K=", best_K)), col=c("black", "red", "blue"), lty=c(1:3), lwd=2)
dev.off()

# SVM and clustering recall comparison
filename <- "svm_hit_rate";
svm_hit_rate <- read.octave(paste("serial/", filename, ".mat", sep=""))
svm_hit_rate <- svm_hit_rate[[1]]
filename <- "bernoulli_hit_rate";
bernoulli_hit_rate <- read.octave(paste("serial/", filename, ".mat", sep=""))
bernoulli_hit_rate <- bernoulli_hit_rate[[1]]
filename <- "baseline_hit_rate_serial";
baseline_hit_rate_serial <- read.octave(paste("serial/", filename, ".mat", sep=""))
baseline_hit_rate_serial <- baseline_hit_rate_serial[[1]]
pdf(file=paste("recall-comparison", ".pdf", sep=""), width=7, height=4)
last <- length(svm_hit_rate[1, ])
plot(2:last, svm_hit_rate[3, 2:last], type="l", col="black", ylim=c(0, 1), xlab="Iteration", ylab="Recall", lty=1, lwd=2)
last <- length(bernoulli_hit_rate[1, ])
lines(2:last, bernoulli_hit_rate[3, 2:last], type="l", col="red", ylim=c(0, 1), lty=2, lwd=2)
last <- length(baseline_hit_rate_serial[best_K, , 3])
lines(2:last, baseline_hit_rate_serial[best_K, 2:last, 3], type="l", col="blue", ylim=c(0, 1), lty=3, lwd=2)
title(main="Recall Comparison", col.main="black", font.main=4)
legend(x="bottomright", legend=c("SVM", "Bernoulli", paste("Clustering, K=", best_K)), col=c("black", "red", "blue"), lty=c(1:3), lwd=2)
dev.off()

# SVM and clustering F-measure comparison
filename <- "svm_hit_rate";
svm_hit_rate <- read.octave(paste("serial/", filename, ".mat", sep=""))
svm_hit_rate <- svm_hit_rate[[1]]
filename <- "bernoulli_hit_rate";
bernoulli_hit_rate <- read.octave(paste("serial/", filename, ".mat", sep=""))
bernoulli_hit_rate <- bernoulli_hit_rate[[1]]
filename <- "baseline_hit_rate_serial";
baseline_hit_rate_serial <- read.octave(paste("serial/", filename, ".mat", sep=""))
baseline_hit_rate_serial <- baseline_hit_rate_serial[[1]]
pdf(file=paste("f-measure-comparison", ".pdf", sep=""), width=7, height=4)
last <- length(svm_hit_rate[1, ])
plot(2:last, svm_hit_rate[4, 2:last], type="l", col="black", ylim=c(0, 1), xlab="Iteration", ylab="F-measure", lty=1, lwd=2)
last <- length(bernoulli_hit_rate[1, ])
lines(2:last, bernoulli_hit_rate[4, 2:last], type="l", col="red", ylim=c(0, 1), lty=2, lwd=2)
last <- length(baseline_hit_rate_serial[best_K, , 4])
lines(2:last, baseline_hit_rate_serial[best_K, 2:last, 4], type="l", col="blue", ylim=c(0, 1), lty=3, lwd=2)
title(main="F-measure Comparison", col.main="black", font.main=4)
legend(x="bottomright", legend=c("SVM", "Bernoulli", paste("Clustering, K=", best_K)), col=c("black", "red", "blue"), lty=c(1:3), lwd=2)
dev.off()

# baseline & SVM accuracy
#filename <- "baseline_accuracy"
#baseline_accuracy <- read.octave(paste("parallel/", filename, ".mat", sep=""))
#baseline_accuracy <- baseline_accuracy[[1]]
#filename <- "svm_accuracy"
#svm_accuracy <- read.octave(paste("parallel/", filename, ".mat", sep=""))
#svm_accuracy <- svm_accuracy[[1]]
#last <- length(baseline_accuracy[1, , 1])
#max_K <- length(baseline_accuracy[1, 1, ])
#pdf(file=paste("accuracies", ".pdf", sep=""), width=7, height=4)
#accuracies <- cbind(baseline_accuracy[1, 2:last, 2:max_K], svm_accuracy[1, 2:last, 2])
#names <- vector()
#for(K in 2:max_K) {
#  names[K-1] <- paste("K=", K)
#}
#names[max_K] <- "SVM"
#boxplot(accuracies, xlab="Method", ylab="Accuracy", names=names)
#title(main=paste("Accuracies (parallel)"), col.main="black", font.main=4)
#dev.off()

# baseline & SVM training
#filename <- "baseline_learning"
#baseline_learning <- read.octave(paste("parallel/", filename, ".mat", sep=""))
#baseline_learning <- baseline_learning[[1]]
#filename <- "svm_learning"
#svm_learning <- read.octave(paste("parallel/", filename, ".mat", sep=""))
#svm_learning <- svm_learning[[1]]
#last <- length(baseline_learning[1, , 1])
#max_K <- length(baseline_learning[1, 1, ])
#pdf(file=paste("learning-durations", ".pdf", sep=""), width=7, height=4)
#accuracies <- cbind(baseline_learning[1, 2:last, 2:max_K], svm_learning[1, 2:last, 2])
#names <- vector()
#for(K in 2:max_K) {
#  names[K-1] <- paste("K=", K)
#}
#names[max_K] <- "SVM"
#boxplot(accuracies, xlab="Method", ylab="Learning durations", names=names)
#title(main=paste("Learning Durations (parallel)"), col.main="black", font.main=4)
#dev.off()

# baseline & SVM prediction
#filename <- "baseline_prediction"
#baseline_prediction <- read.octave(paste("parallel/", filename, ".mat", sep=""))
#baseline_prediction <- baseline_prediction[[1]]
#filename <- "svm_prediction"
#svm_prediction <- read.octave(paste("parallel/", filename, ".mat", sep=""))
#svm_prediction <- svm_prediction[[1]]
#last <- length(baseline_prediction[1, , 1])
#max_K <- length(baseline_prediction[1, 1, ])
#pdf(file=paste("prediction-durations", ".pdf", sep=""), width=7, height=4)
#accuracies <- cbind(baseline_prediction[1, 2:last, 2:max_K], svm_prediction[1, 2:last, 2])
#names <- vector()
#for(K in 2:max_K) {
#  names[K-1] <- paste("K=", K)
#}
#names[max_K] <- "SVM"
#boxplot(accuracies, xlab="Method", ylab="Prediction durations", names=names)
#title(main=paste("Prediction Durations (parallel)"), col.main="black", font.main=4)
#dev.off()


